{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "726775a5-c4b7-4fb4-a36d-ce151bbca008",
   "metadata": {},
   "source": [
    "# SQL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nHere we will be creating an SQL database to store disease signatures from iLINCS!\\n\\nResources: \\n    * http://www.ilincs.org/ilincs/APIinfo\\n'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Here we will be creating an SQL database to store disease signatures from iLINCS!\n",
    "\n",
    "We will filter those signatures which belong to diseases - which are >9,000 signatures\n",
    "from the iLINCS database.\n",
    "\n",
    "Resources: \n",
    "    * http://www.ilincs.org/ilincs/APIinfo\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9528e5a0-747e-479a-8766-7910db900150",
   "metadata": {},
   "source": [
    "## Create Table Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "101377be-c525-46ee-9e60-bc2f3f80f07a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-12-04 14:54:53,827 - INFO - Data loaded successfully.\n",
      "/tmp/ipykernel_134319/2803087972.py:41: DtypeWarning: Columns (0,3,7,8,9,10,15,19,20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_data = pd.read_csv(path_data)\n",
      "2023-12-04 14:54:54,413 - INFO - Data loaded successfully.\n",
      "2023-12-04 14:54:54,459 - INFO - Shape of filtered DataFrame: (1087, 16)\n"
     ]
    }
   ],
   "source": [
    "# 1. Imports, Variables, Functions\n",
    "import psycopg2\n",
    "import pandas as pd\n",
    "import logging\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO, format=\"%(asctime)s - %(levelname)s - %(message)s\"\n",
    ")\n",
    "\n",
    "# SQL variables\n",
    "dbname = \"ilincs\"\n",
    "user = \"ddalton\"\n",
    "password = \"Teclado$$$111\"\n",
    "host = \"localhost\"\n",
    "path_data = \"../../data/iLINCS/datasets.csv\"\n",
    "path_signature = \"../../data/iLINCS/signatures.csv\"\n",
    "table_name = \"datasets\"\n",
    "primary_key = \"experiment\"\n",
    "int_columns = [\"nsamples\"]  # INT columns - rest TEXT\n",
    "drop_table = True\n",
    "\n",
    "\n",
    "# functions\n",
    "def get_disease_datasets():\n",
    "    \"\"\"\n",
    "    Get Disease Datasets\n",
    "    Function to retrieve from those filtered signatures the datasetid\n",
    "\n",
    "    Arguments:\n",
    "\n",
    "    Return:\n",
    "    datasetid: list()\n",
    "        List of unique dataset id's\n",
    "    \"\"\"\n",
    "    path_data = \"../../data/iLINCS/signatures.csv\"\n",
    "    filter_df = lambda df: df[\"libraryid\"] == \"LIB_1\"\n",
    "\n",
    "    # Load Data\n",
    "    try:\n",
    "        df_data = pd.read_csv(path_data)\n",
    "        logging.info(\"Data loaded successfully.\")\n",
    "    except FileNotFoundError:\n",
    "        logging.error(\"Data file not found. Please check the file path.\")\n",
    "        exit()\n",
    "\n",
    "    # filter disease signatures\n",
    "    df_data = df_data[filter_df]\n",
    "\n",
    "    return list(df_data[\"datasetid\"].unique())\n",
    "\n",
    "\n",
    "# 2. Load Data\n",
    "try:\n",
    "    df_data = pd.read_csv(path_data)\n",
    "    logging.info(\"Data loaded successfully.\")\n",
    "except FileNotFoundError:\n",
    "    logging.error(\"Data file not found. Please check the file path.\")\n",
    "    exit()\n",
    "\n",
    "# get unique disease dataset id's\n",
    "unique_datasetid = get_disease_datasets()\n",
    "\n",
    "# filter disease signatures\n",
    "# dataset id's refer to experiment\n",
    "df_data = df_data[df_data[\"experiment\"].isin(unique_datasetid)]\n",
    "\n",
    "# assert we find all dataset id's\n",
    "assert len(unique_datasetid) == len(\n",
    "    df_data[\"experiment\"].unique()\n",
    "), \"Error, Dataset Table does not conain all unique datasetid\"\n",
    "\n",
    "# get max length for each\n",
    "max_lengths = [\n",
    "    max([len(str(n)) for n in df_data[c].to_list()]) for c in df_data.columns\n",
    "]\n",
    "\n",
    "# Convert specified integer columns and handle NaN by replacing with 0\n",
    "for col in int_columns:\n",
    "    df_data[col] = pd.to_numeric(df_data[col], errors=\"coerce\").fillna(0).astype(int)\n",
    "\n",
    "# For other columns, replace NaN with None (which will become NULL in SQL)\n",
    "for col in df_data.columns:\n",
    "    if col not in int_columns:\n",
    "        df_data[col] = df_data[col].where(pd.notnull(df_data[col]), None)\n",
    "\n",
    "# Drop Duplicate for experiment column\n",
    "df_data = df_data.drop_duplicates(subset=\"experiment\", keep=\"first\")\n",
    "\n",
    "logging.info(f\"Shape of filtered DataFrame: {df_data.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "799fc809-f316-4be7-b0aa-ab6cf204b6d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-12-04 14:55:12,807 - INFO - Connected to the database successfully.\n",
      "2023-12-04 14:55:12,813 - INFO - Table datasets created successfully.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table datasets dropped successfully if it existed.\n"
     ]
    }
   ],
   "source": [
    "# 3. Connect with Database\n",
    "try:\n",
    "    conn = psycopg2.connect(dbname=dbname, user=user, password=password, host=host)\n",
    "    logging.info(\"Connected to the database successfully.\")\n",
    "except psycopg2.OperationalError as e:\n",
    "    logging.error(f\"Unable to connect to the database: {e}\")\n",
    "    exit()\n",
    "\n",
    "# 4. Create Cursor Object\n",
    "cursor = conn.cursor()\n",
    "\n",
    "# 5. Check if Table Exists and Delete Data if It Does\n",
    "# Check if the table exists and drop it if it does\n",
    "if drop_table:\n",
    "    try:\n",
    "        cursor.execute(f\"DROP TABLE IF EXISTS {table_name};\")\n",
    "        conn.commit()\n",
    "        print(f\"Table {table_name} dropped successfully if it existed.\")\n",
    "    except psycopg2.Error as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        conn.rollback()\n",
    "\n",
    "\n",
    "cursor.execute(\n",
    "    \"SELECT EXISTS(SELECT * FROM information_schema.tables WHERE table_name=%s)\",\n",
    "    (table_name,),\n",
    ")\n",
    "table_exists = cursor.fetchone()[0]\n",
    "\n",
    "if table_exists:\n",
    "    try:\n",
    "        cursor.execute(f\"DELETE FROM {table_name};\")\n",
    "        conn.commit()\n",
    "        logging.info(f\"Existing data in table {table_name} deleted successfully.\")\n",
    "    except psycopg2.Error as e:\n",
    "        logging.error(f\"An error occurred while deleting data from the table: {e}\")\n",
    "        cursor.close()\n",
    "        conn.close()\n",
    "        exit()\n",
    "else:\n",
    "    # Create table if it does not exist\n",
    "    column_text = \", \".join(\n",
    "        f\"{c} VARCHAR({n + 10})\" if c not in int_columns else f\"{c} INT\"\n",
    "        for c, n in zip(df_data.columns, max_lengths)\n",
    "    )\n",
    "    create_table_query = (\n",
    "        f\"CREATE TABLE {table_name} ({column_text}, PRIMARY KEY({primary_key}));\"\n",
    "    )\n",
    "    try:\n",
    "        cursor.execute(create_table_query)\n",
    "        conn.commit()\n",
    "        logging.info(f\"Table {table_name} created successfully.\")\n",
    "    except psycopg2.Error as e:\n",
    "        logging.error(f\"An error occurred while creating the table: {e}\")\n",
    "        cursor.close()\n",
    "        conn.close()\n",
    "        exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "005fc4a9-b177-484a-9372-1b024c855443",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-12-04 14:55:17,467 - INFO - Data dumped into datasets successfully.\n"
     ]
    }
   ],
   "source": [
    "# 6. Dump Data into Table\n",
    "data_tuples = list(df_data.itertuples(index=False, name=None))\n",
    "insert_query = (\n",
    "    f\"INSERT INTO {table_name} ({', '.join(df_data.columns)}) VALUES (%s\"\n",
    "    + \", %s\" * (len(df_data.columns) - 1)\n",
    "    + \")\"\n",
    ")\n",
    "\n",
    "try:\n",
    "    with conn:\n",
    "        with conn.cursor() as curs:\n",
    "            for record in data_tuples:\n",
    "                try:\n",
    "                    curs.execute(insert_query, record)\n",
    "                except psycopg2.Error as e:\n",
    "                    logging.error(f\"Error inserting record {record}: {e}\")\n",
    "                    # Optionally, you can break the loop after logging the first error\n",
    "                    break\n",
    "    logging.info(f\"Data dumped into {table_name} successfully.\")\n",
    "except psycopg2.Error as e:\n",
    "    logging.error(f\"An error occurred while inserting data into the table: {e}\")\n",
    "    conn.rollback()\n",
    "    exit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5be09a16-2962-49b1-b4ad-f1c23a35051a",
   "metadata": {},
   "source": [
    "## Create Signature Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ec2f24cf-83c8-4522-8b96-7d00b86283bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_135693/106177209.py:28: DtypeWarning: Columns (0,3,7,8,9,10,15,19,20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_data = pd.read_csv(path_data)\n",
      "2023-12-04 14:59:17,169 - INFO - Data loaded successfully.\n",
      "2023-12-04 14:59:17,234 - INFO - Shape of filtered DataFrame: (9097, 30)\n",
      "2023-12-04 14:59:17,241 - INFO - Connected to the database successfully.\n",
      "2023-12-04 14:59:17,248 - INFO - Table signatures created successfully.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table signatures dropped successfully if it existed.\n"
     ]
    }
   ],
   "source": [
    "# 1. Imports, Variables, Functions\n",
    "import psycopg2\n",
    "import pandas as pd\n",
    "import logging\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO, format=\"%(asctime)s - %(levelname)s - %(message)s\"\n",
    ")\n",
    "\n",
    "# SQL variables\n",
    "dbname = \"ilincs\"\n",
    "user = \"ddalton\"\n",
    "password = \"Teclado$$$111\"\n",
    "host = \"localhost\"\n",
    "path_data = \"../../data/iLINCS/signatures.csv\"\n",
    "table_name = \"signatures\"\n",
    "primary_key = \"signatureid\"\n",
    "int_columns = [\"nCtrSamples\", \"nTrtSamples\", \"pubChemID\"]  # INT columns - rest TEXT\n",
    "drop_table = True\n",
    "filter_df = lambda df: df[\"libraryid\"] == \"LIB_1\"\n",
    "reference_table = \"datasets\"\n",
    "reference_key = \"experiment\"\n",
    "foreign_key = \"datasetid\"\n",
    "\n",
    "# 2. Load Data\n",
    "try:\n",
    "    df_data = pd.read_csv(path_data)\n",
    "    logging.info(\"Data loaded successfully.\")\n",
    "except FileNotFoundError:\n",
    "    logging.error(\"Data file not found. Please check the file path.\")\n",
    "    exit()\n",
    "\n",
    "# filter disease signatures\n",
    "# which are libraryid LIB_1\n",
    "df_data = df_data[filter_df]\n",
    "\n",
    "# get max length for each\n",
    "max_lengths = [\n",
    "    max([len(str(n)) for n in df_data[c].to_list()]) for c in df_data.columns\n",
    "]\n",
    "\n",
    "# Convert specified integer columns and handle NaN by replacing with 0\n",
    "for col in int_columns:\n",
    "    df_data[col] = pd.to_numeric(df_data[col], errors=\"coerce\").fillna(0).astype(int)\n",
    "\n",
    "# For other columns, replace NaN with None (which will become NULL in SQL)\n",
    "for col in df_data.columns:\n",
    "    if col not in int_columns:\n",
    "        df_data[col] = df_data[col].where(pd.notnull(df_data[col]), None)\n",
    "\n",
    "logging.info(f\"Shape of filtered DataFrame: {df_data.shape}\")\n",
    "# 3. Connect with Database\n",
    "try:\n",
    "    conn = psycopg2.connect(dbname=dbname, user=user, password=password, host=host)\n",
    "    logging.info(\"Connected to the database successfully.\")\n",
    "except psycopg2.OperationalError as e:\n",
    "    logging.error(f\"Unable to connect to the database: {e}\")\n",
    "    exit()\n",
    "\n",
    "# 4. Create Cursor Object\n",
    "cursor = conn.cursor()\n",
    "\n",
    "# 5. Check if Table Exists and Delete Data if It Does\n",
    "\n",
    "# Check if the table exists and drop it if it does\n",
    "if drop_table:\n",
    "    try:\n",
    "        cursor.execute(f\"DROP TABLE IF EXISTS {table_name};\")\n",
    "        conn.commit()\n",
    "        print(f\"Table {table_name} dropped successfully if it existed.\")\n",
    "    except psycopg2.Error as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "        conn.rollback()\n",
    "\n",
    "\n",
    "cursor.execute(\n",
    "    \"SELECT EXISTS(SELECT * FROM information_schema.tables WHERE table_name=%s)\",\n",
    "    (table_name,),\n",
    ")\n",
    "table_exists = cursor.fetchone()[0]\n",
    "\n",
    "if table_exists:\n",
    "    try:\n",
    "        cursor.execute(f\"DELETE FROM {table_name};\")\n",
    "        conn.commit()\n",
    "        logging.info(f\"Existing data in table {table_name} deleted successfully.\")\n",
    "    except psycopg2.Error as e:\n",
    "        logging.error(f\"An error occurred while deleting data from the table: {e}\")\n",
    "        cursor.close()\n",
    "        conn.close()\n",
    "        exit()\n",
    "else:\n",
    "    # Create table if it does not exist\n",
    "    column_text = \", \".join(\n",
    "        f\"{c} VARCHAR({n + 10})\" if c not in int_columns else f\"{c} INT\"\n",
    "        for c, n in zip(df_data.columns, max_lengths)\n",
    "    )\n",
    "    create_table_query = f\"CREATE TABLE {table_name} ({column_text}, PRIMARY KEY({primary_key}),FOREIGN KEY ({foreign_key}) REFERENCES {reference_table}({reference_key}));\"\n",
    "    try:\n",
    "        cursor.execute(create_table_query)\n",
    "        conn.commit()\n",
    "        logging.info(f\"Table {table_name} created successfully.\")\n",
    "    except psycopg2.Error as e:\n",
    "        logging.error(f\"An error occurred while creating the table: {e}\")\n",
    "        cursor.close()\n",
    "        conn.close()\n",
    "        exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "85d8832a-455a-4a3a-9e0b-6378a73a6bd7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-12-04 14:59:24,120 - INFO - Data dumped into signatures successfully.\n"
     ]
    }
   ],
   "source": [
    "# 6. Dump Data into Table\n",
    "data_tuples = list(df_data.itertuples(index=False, name=None))\n",
    "insert_query = (\n",
    "    f\"INSERT INTO {table_name} ({', '.join(df_data.columns)}) VALUES (%s\"\n",
    "    + \", %s\" * (len(df_data.columns) - 1)\n",
    "    + \")\"\n",
    ")\n",
    "\n",
    "try:\n",
    "    with conn:\n",
    "        with conn.cursor() as curs:\n",
    "            for record in data_tuples:\n",
    "                try:\n",
    "                    curs.execute(insert_query, record)\n",
    "                except psycopg2.Error as e:\n",
    "                    logging.error(f\"Error inserting record {record}: {e}\")\n",
    "                    # Optionally, you can break the loop after logging the first error\n",
    "                    break\n",
    "    logging.info(f\"Data dumped into {table_name} successfully.\")\n",
    "except psycopg2.Error as e:\n",
    "    logging.error(f\"An error occurred while inserting data into the table: {e}\")\n",
    "    conn.rollback()\n",
    "    exit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "991f809d-e8dd-4504-a659-eeb6071e8137",
   "metadata": {},
   "source": [
    "## Create Table Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d5f5675a-0af8-426c-ae0c-945f3a392b0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-11-27 16:06:40,379 - INFO - Data loaded successfully.\n"
     ]
    }
   ],
   "source": [
    "# 1. Imports, Variables, Functions\n",
    "import psycopg2\n",
    "import pandas as pd\n",
    "import logging\n",
    "\n",
    "# Set up logging\n",
    "logging.basicConfig(\n",
    "    level=logging.INFO, format=\"%(asctime)s - %(levelname)s - %(message)s\"\n",
    ")\n",
    "\n",
    "# SQL variables\n",
    "dbname = \"ilincs\"\n",
    "user = \"ddalton\"\n",
    "password = \"Teclado$$$111\"\n",
    "host = \"localhost\"\n",
    "path_data = \"../../data/iLINCS/genes.csv\"\n",
    "table_name = \"genes\"\n",
    "primary_keys = []\n",
    "int_columns = []  # INT columns - rest TEXT\n",
    "drop_table = True\n",
    "foreign_key = \"\"\n",
    "parent_table = \"\"\n",
    "parent_id = \"\"\n",
    "# 2. Load Data\n",
    "try:\n",
    "    df_data = pd.read_csv(path_data)\n",
    "    logging.info(\"Data loaded successfully.\")\n",
    "except FileNotFoundError:\n",
    "    logging.error(\"Data file not found. Please check the file path.\")\n",
    "    exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "10f9085a-eba8-4d7c-a8f4-b4d009202080",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>description</th>\n",
       "      <th>geneid</th>\n",
       "      <th>homologeneid</th>\n",
       "      <th>locustag</th>\n",
       "      <th>maplocation</th>\n",
       "      <th>namenomauth</th>\n",
       "      <th>nomstatus</th>\n",
       "      <th>symbol</th>\n",
       "      <th>symbonomauth</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>taxid</th>\n",
       "      <th>typeofgene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>11167.0</td>\n",
       "      <td>-</td>\n",
       "      <td>19q13.4</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>2</td>\n",
       "      <td>37248.0</td>\n",
       "      <td>-</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>A2M</td>\n",
       "      <td>A2M</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000256069</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>9606</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11</td>\n",
       "      <td>HGNC:11|MIM:108985</td>\n",
       "      <td>atrophia areata, peripapillary chorioretinal d...</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>11p15</td>\n",
       "      <td>atrophia areata, peripapillary chorioretinal d...</td>\n",
       "      <td>0</td>\n",
       "      <td>AA</td>\n",
       "      <td>AA</td>\n",
       "      <td>-</td>\n",
       "      <td>9606</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...</td>\n",
       "      <td>N-acetyltransferase 1 (arylamine N-acetyltrans...</td>\n",
       "      <td>9</td>\n",
       "      <td>37329.0</td>\n",
       "      <td>-</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1 (arylamine N-acetyltrans...</td>\n",
       "      <td>O</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>9606</td>\n",
       "      <td>protein-coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235977</th>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>ribosomal protein S27a pseudogene</td>\n",
       "      <td>105650979</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>LOC105650979</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>10116</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235978</th>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>ribosomal protein S27a pseudogene</td>\n",
       "      <td>105650980</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>LOC105650980</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>10116</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235979</th>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>NLR family, pyrin domain containing 5 pseudogene</td>\n",
       "      <td>105661729</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>LOC105661729</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>10090</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235980</th>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>long intergenic non-coding RNA muscle differen...</td>\n",
       "      <td>105661730</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Linc-md1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>10090</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235981</th>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>long non-coding RNA LncTCF7</td>\n",
       "      <td>105664404</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>LNCTCF7</td>\n",
       "      <td>-</td>\n",
       "      <td>TCONS_00009511</td>\n",
       "      <td>9606</td>\n",
       "      <td>ncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>235982 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chromosome                                            dbxrefs  \\\n",
       "0              19  MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...   \n",
       "1              12  MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...   \n",
       "2              12                HGNC:HGNC:8|Ensembl:ENSG00000256069   \n",
       "3              11                                 HGNC:11|MIM:108985   \n",
       "4               8  MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...   \n",
       "...           ...                                                ...   \n",
       "235977          3                                                  -   \n",
       "235978          7                                                  -   \n",
       "235979         12                                                  -   \n",
       "235980          1                                                  -   \n",
       "235981          5                                                  -   \n",
       "\n",
       "                                              description     geneid  \\\n",
       "0                                  alpha-1-B glycoprotein          1   \n",
       "1                                   alpha-2-macroglobulin          2   \n",
       "2                      alpha-2-macroglobulin pseudogene 1          3   \n",
       "3       atrophia areata, peripapillary chorioretinal d...          8   \n",
       "4       N-acetyltransferase 1 (arylamine N-acetyltrans...          9   \n",
       "...                                                   ...        ...   \n",
       "235977                  ribosomal protein S27a pseudogene  105650979   \n",
       "235978                  ribosomal protein S27a pseudogene  105650980   \n",
       "235979   NLR family, pyrin domain containing 5 pseudogene  105661729   \n",
       "235980  long intergenic non-coding RNA muscle differen...  105661730   \n",
       "235981                        long non-coding RNA LncTCF7  105664404   \n",
       "\n",
       "        homologeneid locustag maplocation  \\\n",
       "0            11167.0        -     19q13.4   \n",
       "1            37248.0        -    12p13.31   \n",
       "2                NaN        -    12p13.31   \n",
       "3                NaN        -       11p15   \n",
       "4            37329.0        -        8p22   \n",
       "...              ...      ...         ...   \n",
       "235977           NaN        -           -   \n",
       "235978           NaN        -           -   \n",
       "235979           NaN        -           -   \n",
       "235980           NaN        -           -   \n",
       "235981           NaN        -           -   \n",
       "\n",
       "                                              namenomauth nomstatus  \\\n",
       "0                                  alpha-1-B glycoprotein         O   \n",
       "1                                   alpha-2-macroglobulin         O   \n",
       "2                      alpha-2-macroglobulin pseudogene 1         O   \n",
       "3       atrophia areata, peripapillary chorioretinal d...         0   \n",
       "4       N-acetyltransferase 1 (arylamine N-acetyltrans...         O   \n",
       "...                                                   ...       ...   \n",
       "235977                                                  -         -   \n",
       "235978                                                  -         -   \n",
       "235979                                                  -         -   \n",
       "235980                                                  -         -   \n",
       "235981                                                  -         -   \n",
       "\n",
       "              symbol symbonomauth                   synonyms  taxid  \\\n",
       "0               A1BG         A1BG       A1B|ABG|GAB|HYST2477   9606   \n",
       "1                A2M          A2M  A2MD|CPAMD5|FWP007|S863-7   9606   \n",
       "2              A2MP1        A2MP1                       A2MP   9606   \n",
       "3                 AA           AA                          -   9606   \n",
       "4               NAT1         NAT1       AAC1|MNAT|NAT-1|NATI   9606   \n",
       "...              ...          ...                        ...    ...   \n",
       "235977  LOC105650979            -                          -  10116   \n",
       "235978  LOC105650980            -                          -  10116   \n",
       "235979  LOC105661729            -                          -  10090   \n",
       "235980      Linc-md1            -                          -  10090   \n",
       "235981       LNCTCF7            -             TCONS_00009511   9606   \n",
       "\n",
       "            typeofgene  \n",
       "0       protein-coding  \n",
       "1       protein-coding  \n",
       "2               pseudo  \n",
       "3              unknown  \n",
       "4       protein-coding  \n",
       "...                ...  \n",
       "235977          pseudo  \n",
       "235978          pseudo  \n",
       "235979          pseudo  \n",
       "235980           ncRNA  \n",
       "235981           ncRNA  \n",
       "\n",
       "[235982 rows x 14 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de2f16b1-2756-4b04-9080-33aa156cac6d",
   "metadata": {},
   "source": [
    "## Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fafed3de-1966-4f52-aaac-bcc0a09055ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connected to the database successfully.\n",
      "     antibodytarget cellline tissue       cid compound concentration  \\\n",
      "0              None     None   None     cID_1     None          None   \n",
      "1              None     None   None    cID_10     None          None   \n",
      "2              None     None   None   cID_100     None          None   \n",
      "3              None     None   None  cID_1000     None          None   \n",
      "4              None     None   None  cID_1001     None          None   \n",
      "...             ...      ...    ...       ...      ...           ...   \n",
      "9092           None     None   None   cID_995     None          None   \n",
      "9093           None     None   None   cID_996     None          None   \n",
      "9094           None     None   None   cID_997     None          None   \n",
      "9095           None     None   None   cID_998     None          None   \n",
      "9096           None     None   None   cID_999     None          None   \n",
      "\n",
      "     concordancetable   datasetid     factor   level1  ... clueiocompound  \\\n",
      "0     concordanceLib1      GDS690  cell.line   Caco-2  ...           None   \n",
      "1     concordanceLib1      GDS852   protocol  control  ...           None   \n",
      "2     concordanceLib1  gdsGDS1454      other       6q  ...           None   \n",
      "3     concordanceLib1  gdsGDS4296  cell.line     A549  ...           None   \n",
      "4     concordanceLib1  gdsGDS4296  cell.line     A549  ...           None   \n",
      "...               ...         ...        ...      ...  ...            ...   \n",
      "9092  concordanceLib1  gdsGDS4296  cell.line     A498  ...           None   \n",
      "9093  concordanceLib1  gdsGDS4296  cell.line     A498  ...           None   \n",
      "9094  concordanceLib1  gdsGDS4296  cell.line     A549  ...           None   \n",
      "9095  concordanceLib1  gdsGDS4296  cell.line     A549  ...           None   \n",
      "9096  concordanceLib1  gdsGDS4296  cell.line     A549  ...           None   \n",
      "\n",
      "     integratedmoas genetargets  time  treatment perturbagenid stitchid  \\\n",
      "0              None        None  None       None          None     None   \n",
      "1              None        None  None       None          None     None   \n",
      "2              None        None  None       None          None     None   \n",
      "3              None        None  None       None          None     None   \n",
      "4              None        None  None       None          None     None   \n",
      "...             ...         ...   ...        ...           ...      ...   \n",
      "9092           None        None  None       None          None     None   \n",
      "9093           None        None  None       None          None     None   \n",
      "9094           None        None  None       None          None     None   \n",
      "9095           None        None  None       None          None     None   \n",
      "9096           None        None  None       None          None     None   \n",
      "\n",
      "     pubchemid is_exemplar pert_type  \n",
      "0            0         NaN      None  \n",
      "1            0         NaN      None  \n",
      "2            0         NaN      None  \n",
      "3            0         NaN      None  \n",
      "4            0         NaN      None  \n",
      "...        ...         ...       ...  \n",
      "9092         0         NaN      None  \n",
      "9093         0         NaN      None  \n",
      "9094         0         NaN      None  \n",
      "9095         0         NaN      None  \n",
      "9096         0         NaN      None  \n",
      "\n",
      "[9097 rows x 30 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_242232/3388944588.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.\n",
      "  df = pd.read_sql(query, conn)\n"
     ]
    }
   ],
   "source": [
    "# SQL variables\n",
    "dbname = \"ilincs\"\n",
    "user = \"ddalton\"\n",
    "password = \"Teclado$$$111\"\n",
    "host = \"localhost\"\n",
    "table_name = \"signatures\"\n",
    "\n",
    "# Connect to the database\n",
    "try:\n",
    "    conn = psycopg2.connect(dbname=dbname, user=user, password=password, host=host)\n",
    "    print(\"Connected to the database successfully.\")\n",
    "except psycopg2.OperationalError as e:\n",
    "    print(f\"Unable to connect to the database: {e}\")\n",
    "    exit()\n",
    "\n",
    "# Query the table\n",
    "try:\n",
    "    query = f\"SELECT * FROM {table_name} LIMIT 10;\"  # Adjust the query as needed\n",
    "    query = f\"SELECT * FROM {table_name} WHERE antibodytarget IS NOT NULL;\"\n",
    "    query = f\"SELECT * FROM {table_name} WHERE experiment IS  NULL;\"\n",
    "    query = f\"SELECT * FROM {table_name}\"\n",
    "    df = pd.read_sql(query, conn)\n",
    "    print(df)\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred while querying the table: {e}\")\n",
    "finally:\n",
    "    conn.close()\n",
    "\n",
    "# The DataFrame 'df' now contains the first 10 rows of the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
